service/lib/dinstaller/storage/guided_proposal.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+# Copyright (c) [2022] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "singleton"
+require "y2storage/storage_manager"
+require "y2storage/guided_proposal"
+require "y2storage/proposal_settings"
+
+module DInstaller
+  module Storage
+    # Backend class to calculate a storage proposal
+    class GuidedProposal
+      include Singleton
+
+      class NoProposalError < StandardError; end
+
+      attr_reader :proposal
+
+      def available_devices
+        disk_analyzer.candidate_disks
+      end
+
+      def candidate_devices
+        raise NoProposalError unless proposal
+
+        proposal.settings.candidate_devices
+      end
+
+      def lvm?
+        raise NoProposalError unless proposal
+
+        proposal.settings.use_lvm
+      end
+
+      def calculate(settings = {})
+        proposal_settings = generate_proposal_settings(settings)
+
+        @proposal = Y2Storage::GuidedProposal.initial(
+          settings:      proposal_settings,
+          devicegraph:   probed_devicegraph,
+          disk_analyzer: disk_analyzer
+        )
+
+        save unless proposal.failed?
+      end
+
+    private
+
+      def generate_proposal_settings(settings)
+        proposal_settings = Y2Storage::ProposalSettings.new_for_current_product
+
+        proposal_settings.use_lvm = settings["LVM"] if settings.keys.include?("LVM")
+
+        if settings.keys.include?("CandidateDevices")
+          proposal_settings.candidate_devices = settings["CandidateDevices"]
```

Comment:
It is probably intention to have here settings based on DBus call convention (that allows that easy signal that just return values that changed), but I do not like much this coupling of DBus property names and backend hash keys.

Reply:
Well, *settings* is the value of a parameter from a dbus method, so I expect it uses dbus properties names. We need some conversion at some point.

Reply:
yeap, I just this that such conversion should be done between in dbus and not in backend as backend should be dbus agnostic. but just my opinion

Reply:
Ah, good point!